corehq/ex-submodules/couchforms/models.py

```diff
@@ -346,7 +346,7 @@ def _meta_to_json(meta):
     def xml_md5(self):
         return hashlib.md5(self.get_xml()).hexdigest()
 
-    def archive(self, user_id=None, rebuild_models=True):
+    def archive(self, user_id=None, trigger_signals=True):
```

Comment:
What made you decide to go with this approach vs removing the signals?  In particular I thought this comment of yours was salient:

>  the reason for this approach is that if we hook anything else to the signal it should get fired, currently its just one rebuild_form_cases though so it does not sound useful.

I agree that `trigger_signals` is a better name for the param with this implementation, since `xform_archived` could do more than just rebuild models (though I suppose you could also rename the signal to prevent that).  However, I prefer the other approach - calling the parameter `rebuild_models` and replacing the `xform_archived.send` call with direct calls to the signal receiver functions.  The reason being that `trigger_signals=False` doesn't convey any information about what is being skipped or why you might want to skip it, so you'll always have to check what signal receivers there are.

Reply:
So the other implementation would look like
```
def archive(self, user_id=None, rebuild_models=True):
    # do archival of the form
    archive_stub.archive_history_updated()
    if rebuild_models:
        remove_data(self)
        rebuild_form_cases(self)
```

Reply:
> The reason being that trigger_signals=False doesn't convey any information about what is being skipped or why you might want to skip it, so you'll always have to check what signal receivers there are.

Isn't that something true wherever you would have signals? you would need to check what is happening in each receiver to know what's happening via signals?
if yes, then this is just about how signals work and not related to changes being made here?

> I thought this comment of yours was salient
Ya. I was concerned if a new signal is added, it would get skipped too and hence i passed the param to the receiver itself originally. For the implementation now and even for the suggestion you gave above, in both cases one would at least look at the method `archive` and see what's happening in it before hooking `form_archived` signal. Adding the param in the receivers like i was doing earlier or calling the methods directly in your suggestion, does make it more safer/explicit but i think one would not miss the signals if they look at the `archive` method.

I think the signals were originally added coz of maintaining couch and sql form models and hence cross app work and so the use of signals.

Reply:
> Isn't that something true wherever you would have signals?

Yes.  Signals aren't meant to know anything about their receivers.  This can be useful for places where you need to coordinate across two different codebases.  For instance, we use Django's utilities for authentication.  If we wanted to add in our own check that the user has access to say a particular domain, we could hook in to a post_authenticate signal to make sure that always happens.  The django authentication library doesn't know or care about our particular use-case, and we're given essentially one-way access.

If this happens in a single codebase, then a signal doesn't add anything in my opinion (except for error handling in the case of `send_robust`).  It makes it look like two pieces of code are decoupled, and it does at least enforce that the knowledge flows only one way, but it does indirectly call every receiver synchronously.  My preference is generally towards making that kind of linkage obvious, particularly if you do need to know what happens in the receivers.

Reply:
(this an esoteric aside about programming style - definitely not a blocking thread)